server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java

```diff
@@ -107,15 +108,33 @@ public final void collectExistingBucket(LeafBucketCollector subCollector, int do
      * Refer to that method for documentation about the merge map.
      */
     public final void mergeBuckets(long[] mergeMap, long newNumBuckets) {
+        UnaryOperator<Long> mergeMapOperator = new UnaryOperator<Long>() {
+            @Override
+            public Long apply(Long bucket) {
+                return mergeMap[Math.toIntExact(bucket)];
+            }
+        };
+
+        mergeBuckets(mergeMapOperator, newNumBuckets);
+    }
+
+    /**
+     * This only tidies up doc counts. Call {@link MergingBucketsDeferringCollector#mergeBuckets(UnaryOperator)} to
```

Comment:
I think it is worth saying what the unary operator does here, that `-1` means throw away and otherwise it is the destination index.